subsys/fs/ext2/ext2_format.c

```diff
@@ -0,0 +1,456 @@
+	bg->bg_used_dirs_count = 2; /* '/' and 'lost+found' */
+
+	if (ext2_write_block(fs, bg_block) < 0) {
+		ret = -EIO;
+		goto out;
+	}
+
+	/* Block bitmap */
+	uint8_t *bbitmap = bbitmap_block->data;
+
+	/* In bitmap we describe blocks starting from s_first_data_block. */
+	set_bitmap_padding(bbitmap, blocks_count - sb->s_first_data_block, cfg);
+	set_bitmap_bits(bbitmap, used_blocks);
+	if (ext2_write_block(fs, bbitmap_block) < 0) {
+		ret = -EIO;
+		goto out;
+	}
+
+	/* Inode bitmap */
+	uint8_t *ibitmap = ibitmap_block->data;
+
+	set_bitmap_padding(ibitmap, inodes_count, cfg);
+	set_bitmap_bits(ibitmap, used_inodes);
+	if (ext2_write_block(fs, ibitmap_block) < 0) {
+		ret = -EIO;
+		goto out;
+	}
+
+	/* Inode table */
+	/* Zero inode table */
+	for (int i = 0; i < itable_blocks; i++) {
+		struct ext2_block *blk = ext2_get_block(fs, itable_block_num + i);
+
+		memset(blk->data, 0, cfg->block_size);
+		rc = ext2_write_block(fs, blk);
+		ext2_drop_block(blk);
+		if (rc < 0) {
+			ret = -EIO;
+			goto out;
+		}
+	}
+
+	struct ext2_disk_inode *in;
+	int inode_offset;
+
+	/* Set inode 2 ('/' directory) */
+	itable_block1 = ext2_get_block(fs, itable_block_num);
+	in = (struct ext2_disk_inode *)itable_block1->data;
+	inode_offset = EXT2_ROOT_INODE - 1;
+	default_directory_inode(&in[inode_offset], 1, cfg);
+
+	in[inode_offset].i_mode = EXT2_DEF_DIR_MODE;
+	in[inode_offset].i_links_count = 3; /* 2 from itself and 1 from child directory */
+	in[inode_offset].i_block[0] = root_dir_blk_num;
+	if (ext2_write_block(fs, itable_block1) < 0) {
+		ret = -EIO;
+		goto out;
+	}
+
+	/* Set inode for 'lost+found' directory */
+	inode_offset = (lost_found_inode - 1) % inodes_per_block; /* We count inodes from 1 */
+
+	LOG_DBG("Inode offset: %d", inode_offset);
+
+	if (inodes_per_block < lost_found_inode) {
+		/* We need to fetch new inode table block */
+		uint32_t block_num = itable_block_num + lost_found_inode / inodes_per_block;
+
+		itable_block2 = ext2_get_block(fs, block_num);
+		in = (struct ext2_disk_inode *)itable_block2->data;
+	}
+
+	default_directory_inode(&in[inode_offset], 1, cfg);
+	in[inode_offset].i_mode = EXT2_DEF_DIR_MODE;
+	in[inode_offset].i_links_count = 2; /* 1 from itself and 1 from parent directory */
+	in[inode_offset].i_block[0] = lost_found_dir_blk_num;
+	if (itable_block2) {
+		if (ext2_write_block(fs, itable_block2) < 0) {
+			ret = -EIO;
+			goto out;
+		}
+	}
+
+	struct ext2_disk_dentry *de;
+	uint32_t current_size;
+
+	/* Contents of '/' directory */
+	LOG_DBG("Root dir blk: %d", root_dir_blk_num);
+	root_dir_blk = ext2_get_block(fs, root_dir_blk_num);
+	if (root_dir_blk == NULL) {
+		ret = ENOMEM;
+		goto out;
+	}
+	memset(root_dir_blk->data, 0, cfg->block_size);
+
+	de = (struct ext2_disk_dentry *)root_dir_blk->data;
+	current_size = 0;
+
+	de->de_inode = 2;
+	de->de_rec_len = sizeof(struct ext2_disk_dentry) + 4;
+	de->de_name_len = 1;
+	de->de_file_type = EXT2_FT_DIR;
+	strcpy(de->de_name, ".");
```

Comment:
I would use `memcpy` here, as you know the length of what is being copied and `strcpy` does not, you do not have to check for 0 `strcpy` does (extra CPU cycles).

Reply:
Also, I am quite that name in node is stored without terminating 0, and `strcpy` does copy 0.

Reply:
I have fixed that (and following comments about directory entries) with common function to fill in the directory entry structure. You are right, names may be stored without terminating 0. I'm using `memcpy` in that function.